src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonIEnumerableDefaultConverter.cs

```diff
@@ -0,0 +1,295 @@
+                            // No need to set PropertyState to TryRead since we're done with this element now.
+                            state.Current.EndElement();
+                        }
+                    }
+
+                    state.Current.ObjectState = StackFrameObjectState.ReadElements;
+                }
+
+                if (state.Current.ObjectState < StackFrameObjectState.EndToken)
+                {
+                    state.Current.ObjectState = StackFrameObjectState.EndToken;
+
+                    // Read the EndObject for $values.
+                    if (state.Current.MetadataId != null)
+                    {
+                        if (!reader.Read())
+                        {
+                            value = default!;
+                            return false;
+                        }
+
+                        if (reader.TokenType != JsonTokenType.EndObject)
+                        {
+                            ThrowHelper.ThrowJsonException_MetadataPreservedArrayInvalidProperty(typeToConvert, reader, ref state);
+                        }
+                    }
+                }
+
+                if (state.Current.ObjectState < StackFrameObjectState.EndTokenValidation)
+                {
+                    if (state.Current.MetadataId != null)
+                    {
+                        if (reader.TokenType != JsonTokenType.EndObject)
+                        {
+                            ThrowHelper.ThrowJsonException_MetadataPreservedArrayInvalidProperty(typeToConvert, reader, ref state);
+                        }
+                    }
+                }
+            }
+
+            ConvertCollection(ref state, options);
+            value = (TCollection)state.Current.ReturnValue!;
+            return true;
+        }
+
+        internal override sealed bool OnTryWrite(Utf8JsonWriter writer, TCollection value, JsonSerializerOptions options, ref WriteStack state)
+        {
+            bool success;
+
+            if (value == null)
+            {
+                writer.WriteNullValue();
+                success = true;
+            }
+            else
+            {
+                bool shouldWritePreservedReferences = options.ReferenceHandling.ShouldWritePreservedReferences();
+
+                if (!state.Current.ProcessedStartToken)
+                {
+                    state.Current.ProcessedStartToken = true;
+
+                    if (!shouldWritePreservedReferences)
+                    {
+                        writer.WriteStartArray();
+                    }
+                    else
+                    {
+                        MetadataPropertyName metadata = JsonSerializer.WriteReferenceForCollection(this, value, ref state, writer);
+                        if (metadata == MetadataPropertyName.Ref)
+                        {
+                            return true;
+                        }
+
+                        state.Current.MetadataPropertyName = metadata;
```

Comment:
This is only remembered to write the `EndObject` token for collections wrapped in an object with metadata; wouldn't be better to set a bool in `state.Current` rather than the metadata?

Reply:
`metadata` is just an enum, so effectively just as fast\good as a separate and new `bool` that would add extra memory and overhead for clearing.

Reply:
I was not talking about a new `bool` but about replacing the existing `MetadataPropertyName` for the bool instead. Also I think you can get rid of using `MetadataPropertyName` at all on the write side if you change `WriteReferenceForObject/Collection` to `bool`.

Reply:
`MetadataPropertyName` is also used for the JsonPath functionality, so the value should always be up-to-date in case there is an exception.

Using a bool instead of `MetadataPropertyName` on the write side is a wash; I think it's more readable to use the enum.

Reply:
The Json path belongs to the read side, isn't it? so no need to worry about it on write.

Reply:
Based on discussions MetadataPropertyName is removed from read side. However it is still used on the write side.